src/Libraries/Microsoft.Extensions.AI/ChatCompletion/StreamingChatCompletionUpdateExtensions.cs

```diff
@@ -0,0 +1,266 @@
+    /// <param name="coalesceContent">
+    /// <see langword="true"/> to attempt to coalesce contiguous streaming update <see cref="AIContent"/> items, where applicable,
+    /// into a single <see cref="AIContent"/>, in order to reduce the number of individual content items that are included in
+    /// manufactured <see cref="ChatMessage"/> instances. When <see langword="false"/>, the content items are kept unaltered.
+    /// A value of <see langword="true"/> does not impact the updates yielded from the enumerable, only the final message added
+    /// to <paramref name="chatMessages"/>.
+    /// </param>
+    /// <returns>
+    /// An enumerable that yields the same updates as <paramref name="updates"/>. It must be enumerated to completion
+    /// or disposed to ensure that the final message is added to <paramref name="chatMessages"/>.
+    /// </returns>
+    /// <remarks>
+    /// <para>
+    /// If any updates were yielded, the method will add a message to <paramref name="chatMessages"/> upon
+    /// completion of iteration, regardless of whether completion occurs due to exhausting the input enumerable,
+    /// due to an exception ocurring, or due to early disposal of the enumerator. Each iteration through the
+    /// returned enumerable may possibly add a message, so it should typically only be enumerated once.
+    /// </para>
+    /// <para>
+    /// At most one <see cref="ChatMessage"/> will be added to <paramref name="chatMessages"/>. If the supplied updates
+    /// represent multiple choices, only one choice will be added.
+    /// </para>
+    /// </remarks>
+    public static IAsyncEnumerable<StreamingChatCompletionUpdate> WithMessageAddedAsync(
+        this IAsyncEnumerable<StreamingChatCompletionUpdate> updates,
+        IList<ChatMessage> chatMessages,
+        bool coalesceContent = true)
+    {
+        _ = Throw.IfNull(updates);
+        _ = Throw.IfNull(chatMessages);
+
+        return ImplAsync(updates, chatMessages, coalesceContent);
+
+        static async IAsyncEnumerable<StreamingChatCompletionUpdate> ImplAsync(
+            IAsyncEnumerable<StreamingChatCompletionUpdate> updates,
+            IList<ChatMessage> chatMessages,
+            bool coalesceContent = true)
+        {
+            List<StreamingChatCompletionUpdate> updatesList = [];
+            try
+            {
+                await foreach (var update in updates.ConfigureAwait(false))
+                {
+                    updatesList.Add(update);
+                    yield return update;
+                }
+            }
+            finally
+            {
+                if (coalesceContent)
+                {
+                    CoalesceTextContent(updatesList);
+                }
+
+                if (ToChatCompletion(updatesList) is { Choices.Count: > 0 } completion)
+                {
+                    chatMessages.Add(completion.Message);
```

Comment:
Appending at this point will work fine in interaction-blocking UIs like console apps because the user likely can't send any more messages until the prior streaming process is completed. But in UIs that allow continuous interaction, where a user can post more messages while a chatbot is replying to an older one (and if this doesn't auto-cancel earlier responses), then we'd end up with out-of-order messages.

If we want to make that case work more naturally, we'd need to do something like inserting a placeholder at the beginning of the process, and then populating it with the message at the end.

We could argue it's not really a valid scenario because we expect that, if you post a further message to the list, you'd also cancel any incomplete response that was still coming. But once we lock to a particular behavior it will be hard to change it in the future, so it may be safer to do the placeholder now.